### DISCUSSION TITLE
Question about chained application in python API.

### DISCUSSION THREAD

**Discussion by jparsert** [Z3Prover/z3#6006] [Q&A]
If I wanted to add the assertion  `(assert (= (* a a) b))` via the python API I would write something like this: 
`
s = z3.Solver()
s.add(a * a == b)
`
Now, consider the following assertion: `(assert (= (* a a a) b))` or `(assert (= (* a) b))`. How would I add them in the Python API?

**Comment by LeventErkok** [comment]
Probably easiest to simply use `Product`:

https://github.com/Z3Prover/z3/blob/5a9b0dd747a2fc26513fcb15181678781e6667a7/src/api/python/z3/z3.py#L8829-L8852

You can also use a python reduce expression to achieve the same, if you're willing to always put the arguments in a list:

```python
$ python
Python 3.9.9 (main, Nov 21 2021, 03:23:44)
[Clang 13.0.0 (clang-1300.0.29.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from z3 import *
>>> import functools
>>> x, y = Ints('x y')
>>> s = Solver()
>>> s.add(functools.reduce(lambda a, b: a*b, [x, y, x, x], 1) == x)
>>> print(s.sexpr())
```

This prints:

```
(declare-fun x () Int)
(declare-fun y () Int)
(assert (= (* 1 x y x x) x))
```

And, tangentially, the stack-overflow forum https://stackoverflow.com/questions/tagged/z3py and https://stackoverflow.com/questions/tagged/z3 are more suitable for these sorts of questions.
